docs/BinSkimRules.md

```diff
@@ -914,3 +914,25 @@ This rule emits CSV data to the console for every compiler/language/version comb
 
 ---
 
+## Rule `BA6001.DisableIncrementalLinking`
+
+### Description
+
+Incremental linking support increases binary size and can reduce runtime performance. Fully optimized release builds should not specify incremental linking.
+
+### Messages
+
+#### `Pass`: Pass
+
+'{0}' was compiled with incremental linking disabled.
```

Comment:
>was [](http://example.com/codeflow?start=6&length=3)

So, how confident are we in your 'release' build detection heuristic? That will dictate whether we use 'is' or 'appears to be' below. My note is to include this distinguishing characteristic in the output. We generally emit this data: 1) why did we fire or not, that is., what specific data is salient to the result, 2) what is the impact, and (for errors) 3) what are the range of fixes. E.g., 

'{0}' is a fully optimized release binary that is not compiled with incremental linking, minimizing binary sizes with a corresponding increase in runtime performance.

Reply:
Moderately but not highly confident.

I tried to ask around for better guidance and there didn't seem to really be much.  It is a hard problem to determine whether the intent was a Release or Debug build just by looking at the final binary.  Being part of the build process would provide firm answers to that but would lose the huge scope of BinSkim where it can try to analyze any valid PE file.

The best suggestion for this, which I have not been able to code yet, was to early exit if linking to the debug C runtime.  That should be very strong evidence of a non-release build.  The opposite is not definitive though because it is common to always use the Release C runtime, even for Debug builds.

All that said, I think that these performance rules will always be warnings instead of blocking errors.  That reduces the potential harm significantly.  If we over-eagerly warn about these then there should hopefully not be any damage.

Reply:
That's a good early exit strategy, detection of debug c runtime, that's a good vote for 'debug' and we should take advantage of it, falling back the remaining heuristics otherwise (definitely take your point that the situation will still be ambiguous).

Reply:
This one is more complicated so saving for the next round of addressing feedback.

Reply:
Actually this was easier than I expected.  The `/MTd` and `/MDd` compiler options are a nice clear signal so I have it early-exit as non-optimized when those are used anywhere.  That check is accurate for all of the binaries I have handy for local testing.